sa/sa.go

```diff
@@ -311,7 +311,7 @@ func (ssa *SQLStorageAuthority) AddPrecertificate(ctx context.Context, req *sapb
 		var row struct {
 			Count int64
 		}
-		err := tx.SelectOne(ctx, &row, "SELECT COUNT(*) as count FROM precertificates WHERE serial=?", serialHex)
+		err := tx.SelectOne(ctx, &row, "SELECT COUNT(*) as count FROM precertificates WHERE serial=? FOR UPDATE", serialHex)
```

Comment:
I don't think that adding explicit locking to this query will help, since the purpose of this query is to check for the presence or absence of certain rows. If MariaDB is like Postgres, then the `FOR UPDATE` would not create any row-level locks if it doesn't find any rows with matching serial numbers. You'd need either something like SERIALIZABLE-level predicate locks (expensive) or a unique index (bad for table partitioning) to catch duplicate serial numbers.

Reply:
Apparently InnoDB has "gap locks" and "next-key locks" which actually can prevent creation of rows by other transactions. https://dev.mysql.com/doc/refman/8.4/en/innodb-locking.html#innodb-gap-locks It looks like these are created by `FOR UPDATE`.

Reply:
I've now taken a different approach, that should still prevent simultaneous UPDATES, but doesn't require the whole transaction to operate at the SERIALIZABLE level. PTAL!